include/versioningCompiler/CompilerImpl/JITCompiler.hpp

```diff
@@ -0,0 +1,152 @@
+
+#include "versioningCompiler/Compiler.hpp"
+#include "versioningCompiler/CompilerImpl/ClangLLVM/FileLogDiagnosticConsumer.hpp"
+#include "versioningCompiler/CompilerImpl/ClangLLVM/LLVMInstanceManager.hpp"
+
+#include "clang/Basic/DiagnosticIDs.h"
+#include "clang/Basic/DiagnosticOptions.h"
+#include "clang/Frontend/CompilerInstance.h"
+
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Target/TargetMachine.h"
+
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ExecutionEngine/ExecutionEngine.h"
+#include "llvm/ExecutionEngine/JITSymbol.h"
+#include "llvm/ExecutionEngine/RTDyldMemoryManager.h"
+#include "llvm/ExecutionEngine/SectionMemoryManager.h"
+#include "llvm/ExecutionEngine/Orc/CompileUtils.h"
+#include "llvm/ExecutionEngine/Orc/IRCompileLayer.h"
+#include "llvm/ExecutionEngine/Orc/Core.h"
+#include "llvm/ExecutionEngine/Orc/RTDyldObjectLinkingLayer.h"
+#include "llvm/IR/DataLayout.h"
+#include "llvm/IR/Mangler.h"
+
+#include <string>
+#include <vector>
+
+namespace vc {
+/** Compiler implementation exploiting the clang-as-a-library and llvm ORC JIT paradigm.
+ * It supports the default llvm  optimizations.
+ */
+    class JITCompiler : public Compiler {
+
+    private:
+        // ORC JIT initialization objects
+        llvm::IntrusiveRefCntPtr<clang::DiagnosticIDs> _diagnosticIDs;
+        llvm::IntrusiveRefCntPtr<clang::DiagnosticOptions> _diagnosticOptions;
+        llvm::IntrusiveRefCntPtr<clang::DiagnosticsEngine> _diagEngine;
+        std::shared_ptr<FileLogDiagnosticConsumer> _diagConsumer;
+        std::shared_ptr<LLVMInstanceManager> _llvmManager;
+        std::unique_ptr<llvm::orc::ExecutionSession> _ES;
+        llvm::orc::JITTargetMachineBuilder _JTMB;
+        llvm::orc::ThreadSafeContext _tsctx;
+        llvm::orc::MangleAndInterner _mangle;
+        llvm::DataLayout _dataLayout;
+        llvm::orc::JITDylib &_mainJD;
+
+/** \brief mutex to regulate exclusive access to static command line options
+ * during optimizer option parsing and processing.
+ */
+        static std::mutex opt_parse_mtx;
+
+    public:
+        // Set of maps used to keep a state of the various versions requesting for JIT functionalities.
+        std::map<std::string, std::string> _modules_map; // source string gets placed here rather than the module. If it is present, generate the module
+        std::map<std::string, llvm::orc::ResourceTrackerSP> _resource_tracker_map;
+        std::map<std::string, bool> _isloaded_map;
+        std::map<std::string, std::unique_ptr<llvm::orc::IRCompileLayer>> _layer_map;
+        std::map<std::string, std::unique_ptr<llvm::orc::RTDyldObjectLinkingLayer>> _obj_map;
+
+
+    public:
+        JITCompiler();
+
+        JITCompiler(const std::string &compilerID,
+                    const std::string &libWorkingDir,
+                    const std::string &log //,
+                    //llvm::orc::JITTargetMachineBuilder targetMachineBuilder
```

Comment:
should I stay or should I go?

Reply:
This should be useful to add cross compile support. So far I haven't been able to integrate it seamingless, due to how llvm orc is made and my lack of knowledge in how does it work in its entirety. Should I focus on cross compilation features?

Reply:
I don't like to keep a constructor parameter commented. Either in or out.

If it is useful, we could possibly keep it as a default-initialised parameter, with default value being `native`-ish.

That being said, I do not expect this library to support cross-compilation out-of-the-box in the JIT compiler.
JIT kind of assumes you are already running on the target architecture, and there set of use cases featuring JIT-cross-compilation is ... limited to be generous.

Reply:
I'd go with out.
[Here's why](https://github.com/skeru/libVersioningCompiler/pull/3/files#diff-b30d4a80eeb2d09e85cbb5adfa858ca07d880aca7628c5072860fa618d7352f9R80).

That parameter cannot be kept as such due to how modern orc works. To have a JITTargetMachineBuilder, you need to get a TargetTriple (getTargetTriple), which is being supplied (in this implementation ) by an ExecutorProcessControl, which is contained in an ExecutionSession, which should be generated (I guess) from a SelfExecutorProcessControl (the main process). At least right now this is how this is being done, I don't know how to simplify that section of code honestly